staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -113,12 +113,29 @@ type CloudNodeController struct {
 }
 
 // NewCloudNodeController creates a CloudNodeController object
+//
+//logcheck:context // NewCloudNodeControllerWithContext should be used instead because record.NewBroadcaster is called and works better when a context is supplied (contextual logging, cancellation).
 func NewCloudNodeController(
 	nodeInformer coreinformers.NodeInformer,
 	kubeClient clientset.Interface,
 	cloud cloudprovider.Interface,
 	nodeStatusUpdateFrequency time.Duration,
 	workerCount int32) (*CloudNodeController, error) {
+	ctx := context.Background()
+	return NewCloudNodeControllerWithContext(ctx, nodeInformer, kubeClient, cloud, nodeStatusUpdateFrequency, workerCount)
+}
+
+// NewCloudNodeController creates a CloudNodeController object
+func NewCloudNodeControllerWithContext(
```

Comment:
https://github.com/kubernetes/kubernetes/blob/aed92f4a6e1fb03951b8193470ce7e58a48331c0/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L573 is called by an event handler and wants to emit events. I'm not sure whether that's good or bad - it is what it is.

But why is it necessary to update the node controller? https://github.com/kubernetes/kubernetes/blob/aed92f4a6e1fb03951b8193470ce7e58a48331c0/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L313-L323 is the kind of trivial event handler that I would expect from a controller. It doesn't use the event recorder, so creating that in `Run` should be fine.

Reply:
>According to https://github.com/kubernetes/kubernetes/issues/128121, the problem is that informers are active and try to emit events even though the controller isn't running (yet).

>Should event handlers really emit events? If yes, then this is the right fix. Unfortunately it makes the API more complicated, something that https://github.com/kubernetes/kubernetes/commit/50c12437604b0cd5a73514389409fc2fde8b91bd was trying to avoid.

@pohly I think it is an old bug but it is detected by https://github.com/kubernetes/kubernetes/pull/122145

https://github.com/carlory/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L183

If the init fails, we shouldn't register event handlers. It will cause a crash since https://github.com/kubernetes/kubernetes/pull/122145 is merged.

There's a memory leak in the code. the queue has no worker to consume the event.

Reply:
@pohly agree with @carlory , don't worry, I think that change only surfaced the bug

Reply:
> I was wondering about that when the description said "service-lb-controller is instantiated but not started".

sorry for that, I update the description of this PR.

Reply:
> Should event handlers really emit events? If yes, then this is the right fix

I think the event handlers should not emit events, that is noisy and not providing the right signal to cluster administrators ,